framework/src/media/OutputHandler.cpp

```diff
@@ -0,0 +1,348 @@
+			// Yes... Push data to encode stream as much as possible.
+			size_t pushed = encoder->pushData(buf + wlen, size - wlen);
+			if (pushed == 0) {
+				meddbg("Can not push any data! Error occurred during encoding!\n");
+				break;
+			}
+			wlen += pushed;
+
+			while (1) {
+				// Encoded data size is usually smaller than origin PCM data size,
+				// So we can reuse 'wlen' bytes free space in 'buf'.
+				// Process encoding and get encoded data.
+				size_t ret = wlen;
+				if (!encoder->getFrame(buf, &ret)) {
+					// Normal case, break and continue to push more PCM data
+					break;
+				}
+
+				// Write encoded data to output stream.
+				ssize_t written = writeToStreamBuffer(buf, ret);
+				medvdbg("written size: %d\n", written);
+				if (written != (ssize_t)ret) {
+					meddbg("Can not write all!\n");
+					break;
+				}
+			}
+		} else {
+			// No... Write original PCM data to output stream.
+			wlen += writeToStreamBuffer(buf + wlen, size - wlen);
+			medvdbg("written size : %d\n", wlen);
+			break;
+		}
+	}
+
+	medvdbg("OutputHandler::write(), written %u\n", size);
+	return (ssize_t)wlen;
+}
+
+bool OutputHandler::start()
+{
+	medvdbg("OutputHandler::start()\n");
+	if (!mOutputDataSource->isPrepare()) {
+		return false;
+	}
+
+	createWorker();
+	return true;
+}
+
+bool OutputHandler::stop()
+{
+	medvdbg("OutputHandler::stop()\n");
+	flush();
+
+	destroyWorker();
+	return true;
+}
+
+void OutputHandler::flush()
+{
+	medvdbg("OutputHandler::flush() enter\n");
+	mIsFlushing = true;
+
+	// Worker may be sleeping, wake it up!
+	wakenWorker();
+
+	while (mIsWorkerAlive) {
+		if (mBufferReader->sizeOfData() == 0) {
+			// Flushed
+			break;
+		}
+
+		usleep(20 * 1000);
+	}
+
+	mIsFlushing = false;
+	medvdbg("OutputDataSource::flush() exit\n");
+}
+
+void OutputHandler::createWorker()
+{
+	medvdbg("OutputHandler::createWorker()\n");
+	if (mStreamBuffer && !mIsWorkerAlive) {
+		mStreamBuffer->reset();
+		mIsFlushing = false;
+		mIsWorkerAlive = true;
+
+		long stackSize = CONFIG_OUTPUT_DATASOURCE_STACKSIZE;
+		pthread_attr_t attr;
+		pthread_attr_init(&attr);
+		pthread_attr_setstacksize(&attr, stackSize);
+		int ret = pthread_create(&mWorker, &attr, static_cast<pthread_startroutine_t>(OutputHandler::workerMain), this);
+		if (ret != OK) {
+			meddbg("Fail to create DataSourceWorker thread, return value : %d\n", ret);
+			mIsWorkerAlive = false;
+			return;
+		}
+		pthread_setname_np(mWorker, "OutputHandler");
+	}
+}
+
+void OutputHandler::destroyWorker()
+{
+	medvdbg("OutputHandler::destoryWorker()\n");
+	if (mIsWorkerAlive) {
+		// Setup flag,
+		mIsWorkerAlive = false;
+
+		// Worker may be blocked in buffer reading.
+		mBufferWriter->setEndOfStream();
+
+		// Wake worker up,
+		wakenWorker();
+
+		// Join thread.
+		pthread_join(mWorker, NULL);
+	}
+}
+
+void *OutputHandler::workerMain(void *arg)
+{
+	medvdbg("OutputHandler::workerMain()\n");
+	auto stream = static_cast<OutputHandler *>(arg);
+	auto worker = stream->mOutputDataSource;
```

Comment:
Don't we need to check whether "stream" is valid..?

Reply:
Actually, `workerMain` is used only once, and at that place, arg is always not nullptr.
So there is no need to check, but I added it.